File: ios/BatteryHolder/Models/FirmwareImage.swift
import Foundation

enum FirmwareChannel: String, Codable {
    case stable, beta
}

/// A firmware build from the AWS catalog.
struct FirmwareImage: Identifiable, Codable, Equatable {
    var buildId: String
    var boardId: String
    var version: String
    var channel: FirmwareChannel
    var sizeBytes: Int
    var sha256: String
    var releaseNotes: String
    var createdAt: Date
    /// Presigned download URL (populated by the detail endpoint).
    var downloadUrl: URL?

    var id: String { buildId }

    var sizeDisplay: String {
        ByteCountFormatter.string(fromByteCount: Int64(sizeBytes), countStyle: .file)
    }
}

/// Progress of an over-the-air flash, published by `FirmwareFlasher`.
struct FlashProgress: Equatable {
    enum Phase: Equatable {
        case idle
        case preparing
        case uploading
        case verifying
        case rebooting
        case done
        case failed(String)
    }
    var phase: Phase = .idle
    /// 0...1 for the upload portion.
    var fraction: Double = 0
    var message: String = ""

    var isActive: Bool {
        switch phase {
        case .idle, .done, .failed: return false
        default: return true
        }
    }
}

File: android_app/lib/models/firmware_image.dart
enum FirmwareChannel {
  stable,
  beta;

  static FirmwareChannel fromJson(String raw) => FirmwareChannel.values
      .firstWhere((c) => c.name == raw, orElse: () => FirmwareChannel.stable);
}

/// A firmware build from the AWS catalog.
class FirmwareImage {
  final String buildId;
  final String boardId;
  final String version;
  final FirmwareChannel channel;
  final int sizeBytes;
  final String sha256;
  final String releaseNotes;
  final DateTime createdAt;

  /// Presigned download URL (populated by the detail endpoint).
  final Uri? downloadUrl;

  const FirmwareImage({
    required this.buildId,
    required this.boardId,
    required this.version,
    required this.channel,
    required this.sizeBytes,
    required this.sha256,
    required this.releaseNotes,
    required this.createdAt,
    this.downloadUrl,
  });

  String get id => buildId;

  /// Matches iOS `ByteCountFormatter(countStyle: .file)`: decimal units,
  /// no decimals below MB.
  String get sizeDisplay {
    const units = ['bytes', 'KB', 'MB', 'GB', 'TB'];
    var value = sizeBytes.toDouble();
    var unit = 0;
    while (value >= 1000 && unit < units.length - 1) {
      value /= 1000;
      unit++;
    }
    if (unit == 0) return '$sizeBytes bytes';
    final digits = unit <= 1 ? 0 : 1;
    return '${value.toStringAsFixed(digits)} ${units[unit]}';
  }

  factory FirmwareImage.fromJson(Map<String, dynamic> json) => FirmwareImage(
        buildId: json['buildId'] as String,
        boardId: json['boardId'] as String,
        version: json['version'] as String,
        channel: FirmwareChannel.fromJson(json['channel'] as String),
        sizeBytes: json['sizeBytes'] as int,
        sha256: json['sha256'] as String,
        releaseNotes: json['releaseNotes'] as String? ?? '',
        createdAt: DateTime.parse(json['createdAt'] as String),
        downloadUrl: json['downloadUrl'] == null
            ? null
            : Uri.parse(json['downloadUrl'] as String),
      );
}

/// Phase of an over-the-air flash.
enum FlashPhase { idle, preparing, uploading, verifying, rebooting, done, failed }

/// Progress of an over-the-air flash, published by `FirmwareFlasher`.
class FlashProgress {
  final FlashPhase phase;

  /// 0...1 for the upload portion.
  final double fraction;
  final String message;

  /// Failure text when [phase] is [FlashPhase.failed].
  final String? failureMessage;

  const FlashProgress({
    this.phase = FlashPhase.idle,
    this.fraction = 0,
    this.message = '',
    this.failureMessage,
  });

  bool get isActive => switch (phase) {
        FlashPhase.idle || FlashPhase.done || FlashPhase.failed => false,
        _ => true,
      };

  FlashProgress copyWith({
    FlashPhase? phase,
    double? fraction,
    String? message,
    String? failureMessage,
  }) =>
      FlashProgress(
        phase: phase ?? this.phase,
        fraction: fraction ?? this.fraction,
        message: message ?? this.message,
        failureMessage: failureMessage ?? this.failureMessage,
      );
}
